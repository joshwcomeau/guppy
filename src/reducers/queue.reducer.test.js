import reducer, {
  getPackageJsonLockedForProjectId,
  getNextActionForProjectId,
} from './queue.reducer';
import {
  QUEUE_DEPENDENCY_INSTALL,
  QUEUE_DEPENDENCY_UNINSTALL,
  START_NEXT_ACTION_IN_QUEUE,
} from '../actions';

describe('queue reducer', () => {
  it('should return initial state', () => {
    expect(reducer(undefined, {})).toEqual({});
  });

  it(`should handle ${START_NEXT_ACTION_IN_QUEUE} for queue with next action`, () => {
    const prevState = {
      foo: [
        { action: 'install', dependencies: [{ name: 'redux' }] },
        { action: 'uninstall', dependencies: [{ name: 'react-redux' }] },
      ],
    };

    const action = {
      type: START_NEXT_ACTION_IN_QUEUE,
      projectId: 'foo',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${START_NEXT_ACTION_IN_QUEUE} for queue with no more actions`, () => {
    const prevState = {
      foo: [{ action: 'install', dependencies: [{ name: 'redux' }] }],
    };

    const action = {
      type: START_NEXT_ACTION_IN_QUEUE,
      projectId: 'foo',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${QUEUE_DEPENDENCY_INSTALL} for new dependency on empty queue`, () => {
    const prevState = {};

    const action = {
      type: QUEUE_DEPENDENCY_INSTALL,
      projectId: 'foo',
      name: 'redux',
      version: '3.2',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${QUEUE_DEPENDENCY_INSTALL} for new dependency on existing queue`, () => {
    const prevState = {
      foo: [
        {
          action: 'install',
          dependencies: [{ name: 'react-redux' }],
        },
      ],
    };

    const action = {
      type: QUEUE_DEPENDENCY_INSTALL,
      projectId: 'foo',
      name: 'redux',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${QUEUE_DEPENDENCY_INSTALL} for updating dependency`, () => {
    const prevState = {};

    const action = {
      type: QUEUE_DEPENDENCY_INSTALL,
      projectId: 'foo',
      name: 'redux',
      version: '3.3',
      updating: true,
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${QUEUE_DEPENDENCY_UNINSTALL}`, () => {
    const prevState = {};

    const action = {
      type: QUEUE_DEPENDENCY_UNINSTALL,
      projectId: 'foo',
      name: 'redux',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${QUEUE_DEPENDENCY_INSTALL} for mixed existing queue`, () => {
    const prevState = {
      foo: [
        { action: 'install', dependencies: [{ name: 'react-redux' }] },
        { action: 'uninstall', dependencies: [{ name: 'redux' }] },
      ],
    };

    const action = {
      type: QUEUE_DEPENDENCY_INSTALL,
      projectId: 'foo',
      name: 'lodash',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  it(`should handle ${QUEUE_DEPENDENCY_UNINSTALL} for mixed existing queue`, () => {
    const prevState = {
      foo: [
        { action: 'install', dependencies: [{ name: 'react-redux' }] },
        { action: 'uninstall', dependencies: [{ name: 'redux' }] },
      ],
    };

    const action = {
      type: QUEUE_DEPENDENCY_UNINSTALL,
      projectId: 'foo',
      name: 'lodash',
    };

    expect(reducer(prevState, action)).toMatchSnapshot();
  });

  describe('getPackageJsonLockedForProjectId', () => {
    it('should return false when no dependencies are queued', () => {
      const state = {
        queue: {},
      };

      const projectId = 'foo';

      expect(getPackageJsonLockedForProjectId(state, projectId)).toBe(false);
    });

    it('should return true when dependencies are queued', () => {
      const state = {
        queue: {
          foo: [
            {
              action: 'install',
              dependencies: [{ name: 'redux' }],
            },
          ],
        },
      };

      const projectId = 'foo';

      expect(getPackageJsonLockedForProjectId(state, projectId)).toBe(true);
    });
  });

  describe('getNextActionForProjectId', () => {
    it('should return next action when one is present', () => {
      const state = {
        queue: {
          foo: [{ action: 'install', dependencies: [{ name: 'redux' }] }],
        },
      };

      const projectId = 'foo';

      expect(getNextActionForProjectId(state, projectId)).toMatchSnapshot();
    });

    it('should return undefined when no actions are present', () => {
      const state = {
        queue: {},
      };

      const projectId = 'foo';

      expect(getNextActionForProjectId(state, projectId)).toBe(undefined);
    });
  });
});
