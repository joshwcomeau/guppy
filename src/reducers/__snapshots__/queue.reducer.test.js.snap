// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queue reducer should handle QUEUE_DEPENDENCY_INSTALL for mixed existing queue 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "install",
      "dependencies": Array [
        Object {
          "name": "react-redux",
        },
        Object {
          "name": "lodash",
          "updating": undefined,
          "version": undefined,
        },
      ],
    },
    Object {
      "action": "uninstall",
      "dependencies": Array [
        Object {
          "name": "redux",
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle QUEUE_DEPENDENCY_INSTALL for new dependency on empty queue 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "install",
      "dependencies": Array [
        Object {
          "name": "redux",
          "updating": undefined,
          "version": "3.2",
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle QUEUE_DEPENDENCY_INSTALL for new dependency on existing queue 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "install",
      "dependencies": Array [
        Object {
          "name": "react-redux",
        },
        Object {
          "name": "redux",
          "updating": undefined,
          "version": undefined,
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle QUEUE_DEPENDENCY_INSTALL for updating dependency 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "install",
      "dependencies": Array [
        Object {
          "name": "redux",
          "updating": true,
          "version": "3.3",
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle QUEUE_DEPENDENCY_UNINSTALL 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "uninstall",
      "dependencies": Array [
        Object {
          "name": "redux",
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle QUEUE_DEPENDENCY_UNINSTALL for mixed existing queue 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "install",
      "dependencies": Array [
        Object {
          "name": "react-redux",
        },
      ],
    },
    Object {
      "action": "uninstall",
      "dependencies": Array [
        Object {
          "name": "redux",
        },
        Object {
          "name": "lodash",
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle START_NEXT_ACTION_IN_QUEUE for queue with next action 1`] = `
Object {
  "foo": Array [
    Object {
      "action": "uninstall",
      "dependencies": Array [
        Object {
          "name": "react-redux",
        },
      ],
    },
  ],
}
`;

exports[`queue reducer should handle START_NEXT_ACTION_IN_QUEUE for queue with no more actions 1`] = `Object {}`;
